File: hooks/block-scrape-interval.py
#!/usr/bin/env python3
"""PostToolUse hook (Edit|Write): enforce scrape interval >= 1.0s.

Files under scrape/ directories must respect the minimum request interval
defined in config/common.toml [scrape.manners].  This hook checks:

- TOML: ``interval`` key must exist and be >= 1.0s
- Python: AST-based sleep value check
- Rust / JS / TS / Dart: regex-based sleep value + HTTP call check
"""

import ast
import json
import os
import re
import sys
import tomllib

_MIN_INTERVAL = 1.0

_NOQA_RE = re.compile(r"(?:#|//)\s*noqa:\s*scrape-interval")


# -- Language definitions for regex-based checks --

_SleepPattern = tuple[re.Pattern[str], int, str]  # (regex, capture group, "s"|"ms")


class _LangDef:
    __slots__ = ("http_re", "sleep_patterns")

    def __init__(
        self,
        http_re: re.Pattern[str],
        sleep_patterns: list[_SleepPattern],
    ) -> None:
        self.http_re = http_re
        self.sleep_patterns = sleep_patterns


_PYTHON_HTTP_RE = re.compile(
    r"\b(?:requests|httpx|session|client)\."
    r"(?:get|post|put|patch|delete|head|options|request|send)\b"
    r"|\baiohttp\.ClientSession\b"
    r"|\bpage\.goto\b"
    r"|\burllib\.request\.urlopen\b"
)

_RUST = _LangDef(
    http_re=re.compile(
        r"\breqwest::(?:get|Client)\b"
        r"|\bclient\.(?:get|post|put|patch|delete|head|request)\("
        r"|\.send\(\)\.await\b"
        r"|\bhyper::Client\b"
        r"|\bsurf::(?:get|post|put|delete)\b"
    ),
    sleep_patterns=[
        (re.compile(r"Duration::from_secs\(\s*(\d+)"), 1, "s"),
        (re.compile(r"Duration::from_secs_f(?:32|64)\(\s*([\d.]+)"), 1, "s"),
        (re.compile(r"Duration::from_millis\(\s*(\d+)"), 1, "ms"),
    ],
)

_JS_TS = _LangDef(
    http_re=re.compile(
        r"\bfetch\("
        r"|\baxios\.(?:get|post|put|patch|delete|head|request)\("
        r"|\bgot\("
        r"|\bpage\.goto\("
        r"|\bky\.(?:get|post|put|patch|delete|head)\("
    ),
    sleep_patterns=[
        (re.compile(r"setTimeout\([^,]+,\s*(\d+)"), 1, "ms"),
        (re.compile(r"(?:sleep|delay)\(\s*(\d+)"), 1, "ms"),
        (re.compile(r"waitForTimeout\(\s*(\d+)"), 1, "ms"),
    ],
)

_DART = _LangDef(
    http_re=re.compile(
        r"\bhttp\.(?:get|post|put|patch|delete|head|read)\("
        r"|\bclient\.(?:get|post|put|patch|delete|head|send)\("
        r"|\bDio\(\)"
    ),
    sleep_patterns=[
        (re.compile(r"Duration\(\s*seconds:\s*(\d+)"), 1, "s"),
        (re.compile(r"Duration\(\s*milliseconds:\s*(\d+)"), 1, "ms"),
    ],
)

_EXT_TO_LANG: dict[str, _LangDef] = {
    ".rs": _RUST,
    ".js": _JS_TS,
    ".ts": _JS_TS,
    ".dart": _DART,
}


def _is_scrape_file(file_path: str) -> bool:
    """Return True if file is under a scrape/ directory."""
    norm = file_path.replace("\\", "/")
    return "/scrape/" in norm or norm.startswith("scrape/")


def _is_test_file(file_path: str) -> bool:
    norm = file_path.replace("\\", "/")
    return "/tests/" in norm or os.path.basename(file_path).startswith("test_")


def _parse_interval_str(value: str) -> float | None:
    """Parse interval strings like '1.0s', '500ms', '2s' into seconds."""
    value = value.strip()
    if value.endswith("ms"):
        try:
            return float(value[:-2]) / 1000.0
        except ValueError:
            return None
    if value.endswith("s"):
        try:
            return float(value[:-1])
        except ValueError:
            return None
    try:
        return float(value)
    except ValueError:
        return None


# -- TOML checks --


def _check_toml(file_path: str) -> str | None:
    """Check a TOML file for interval compliance.  Returns reason or None."""
    with open(file_path, "rb") as f:
        data = tomllib.load(f)

    interval_value = _find_interval_in_dict(data)
    if interval_value is None:
        return (
            f"{os.path.basename(file_path)}: interval キーが定義されていません。\n"
            f"scrape設定には interval >= {_MIN_INTERVAL}s を含めてください。"
        )

    seconds = _resolve_interval(interval_value)
    if seconds is None:
        return (
            f"{os.path.basename(file_path)}: interval の値を解析できません: {interval_value!r}"
        )

    if seconds < _MIN_INTERVAL:
        return (
            f"{os.path.basename(file_path)}: interval = {interval_value!r} ({seconds}s) "
            f"は最低 {_MIN_INTERVAL}s 未満です。"
        )

    return None


def _find_interval_in_dict(data: dict) -> object | None:
    """Recursively search for an 'interval' key in nested dicts."""
    if "interval" in data:
        return data["interval"]
    for v in data.values():
        if isinstance(v, dict):
            result = _find_interval_in_dict(v)
            if result is not None:
                return result
    return None


def _resolve_interval(value: object) -> float | None:
    """Convert an interval value (str or number) to seconds."""
    if isinstance(value, (int, float)):
        return float(value)
    if isinstance(value, str):
        return _parse_interval_str(value)
    return None


# -- Regex-based checks (Rust / JS / TS / Dart) --


def _check_regex_lang(file_path: str, lang: _LangDef) -> str | None:
    """Check a source file using regex patterns.  Returns reason or None."""
    with open(file_path, encoding="utf-8") as f:
        source = f.read()

    lines = source.splitlines()
    basename = os.path.basename(file_path)

    low_sleeps = _find_low_sleeps_regex(lines, lang.sleep_patterns)
    if low_sleeps:
        details = "\n".join(f"  L{ln}: {text}" for ln, text in low_sleeps)
        return (
            f"{basename}: sleep/delay の値が {_MIN_INTERVAL}s 未満です。\n"
            f"{details}\n"
            f"リクエスト間隔を {_MIN_INTERVAL}s 以上にしてください "
            f"(// noqa: scrape-interval で除外可)。"
        )

    has_http = _has_http_regex(lines, lang.http_re)
    if has_http and not _has_adequate_sleep_regex(lines, lang.sleep_patterns):
        return (
            f"{basename}: HTTP呼び出しがありますが "
            f"sleep >= {_MIN_INTERVAL}s が見つかりません。\n"
            f"リクエスト間隔を確保する sleep/delay を追加してください "
            f"(// noqa: scrape-interval で除外可)。"
        )

    return None


def _find_low_sleeps_regex(
    lines: list[str], patterns: list[_SleepPattern]
) -> list[tuple[int, str]]:
    """Return (lineno, matched_text) for sleep calls below threshold."""
    violations: list[tuple[int, str]] = []
    for lineno, line in enumerate(lines, 1):
        if _NOQA_RE.search(line):
            continue
        for pat, group, unit in patterns:
            m = pat.search(line)
            if m is None:
                continue
            try:
                value = float(m.group(group))
            except (ValueError, IndexError):
                continue
            seconds = value / 1000.0 if unit == "ms" else value
            if seconds < _MIN_INTERVAL:
                violations.append((lineno, line.strip()))
                break
    return violations


def _has_http_regex(lines: list[str], http_re: re.Pattern[str]) -> bool:
    for line in lines:
        if _NOQA_RE.search(line):
            continue
        if http_re.search(line):
            return True
    return False


def _has_adequate_sleep_regex(
    lines: list[str], patterns: list[_SleepPattern]
) -> bool:
    for line in lines:
        for pat, group, unit in patterns:
            m = pat.search(line)
            if m is None:
                continue
            try:
                value = float(m.group(group))
            except (ValueError, IndexError):
                continue
            seconds = value / 1000.0 if unit == "ms" else value
            if seconds >= _MIN_INTERVAL:
                return True
    return False


# -- Python checks --


def _check_python(file_path: str) -> str | None:
    """Check a Python file for interval compliance.  Returns reason or None."""
    with open(file_path, encoding="utf-8") as f:
        source = f.read()

    lines = source.splitlines()

    # Check 1: sleep calls with value < _MIN_INTERVAL
    low_sleeps = _find_low_sleep_calls(source, lines)
    if low_sleeps:
        details = "\n".join(
            f"  L{lineno}: {call}({value})" for lineno, call, value in low_sleeps
        )
        return (
            f"{os.path.basename(file_path)}: sleep の値が {_MIN_INTERVAL}s 未満です。\n"
            f"{details}\n"
            f"リクエスト間隔を {_MIN_INTERVAL}s 以上にしてください "
            f"(# noqa: scrape-interval で除外可)。"
        )

    # Check 2: HTTP calls present but no adequate sleep anywhere
    has_http = _has_http_calls(source, lines)
    if has_http and not _has_adequate_sleep(source):
        return (
            f"{os.path.basename(file_path)}: HTTP呼び出しがありますが "
            f"sleep >= {_MIN_INTERVAL}s が見つかりません。\n"
            f"リクエスト間隔を確保する sleep を追加してください "
            f"(# noqa: scrape-interval で除外可)。"
        )

    return None


def _find_low_sleep_calls(
    source: str, lines: list[str]
) -> list[tuple[int, str, float]]:
    """Return (lineno, func_name, value) for sleep calls with value < min."""
    try:
        tree = ast.parse(source)
    except SyntaxError:
        return []

    violations: list[tuple[int, str, float]] = []
    for node in ast.walk(tree):
        if not isinstance(node, ast.Call):
            continue
        func_name = _get_sleep_func_name(node.func)
        if func_name is None:
            continue
        if not node.args:
            continue
        arg = node.args[0]
        if not isinstance(arg, ast.Constant) or not isinstance(arg.value, (int, float)):
            continue
        if isinstance(arg.value, bool):
            continue
        value = float(arg.value)
        if value >= _MIN_INTERVAL:
            continue
        line = lines[node.lineno - 1]
        if _NOQA_RE.search(line):
            continue
        violations.append((node.lineno, func_name, value))

    return violations


def _get_sleep_func_name(func: ast.expr) -> str | None:
    """Return 'time.sleep' or 'asyncio.sleep' if the call matches, else None."""
    if isinstance(func, ast.Attribute) and func.attr == "sleep":
        if isinstance(func.value, ast.Name) and func.value.id in ("time", "asyncio"):
            return f"{func.value.id}.sleep"
    if isinstance(func, ast.Name) and func.id == "sleep":
        return "sleep"
    return None


def _has_http_calls(source: str, lines: list[str]) -> bool:
    """Return True if the source contains HTTP request patterns."""
    for line in lines:
        if _NOQA_RE.search(line):
            continue
        if _PYTHON_HTTP_RE.search(line):
            return True
    return False


def _has_adequate_sleep(source: str) -> bool:
    """Return True if at least one sleep >= _MIN_INTERVAL exists."""
    try:
        tree = ast.parse(source)
    except SyntaxError:
        return False
    for node in ast.walk(tree):
        if not isinstance(node, ast.Call):
            continue
        if _get_sleep_func_name(node.func) is None:
            continue
        if not node.args:
            continue
        arg = node.args[0]
        if isinstance(arg, ast.Constant) and isinstance(arg.value, (int, float)):
            if not isinstance(arg.value, bool) and float(arg.value) >= _MIN_INTERVAL:
                return True
    return False


# -- Entry point --


def main() -> None:
    data = json.load(sys.stdin)
    tool_input: dict[str, str] = data.get("tool_input", {})
    file_path: str = tool_input.get("file_path", "") or tool_input.get("path", "")

    if not file_path or not _is_scrape_file(file_path):
        return

    if _is_test_file(file_path):
        return

    if not os.path.isfile(file_path):
        return

    reason: str | None = None
    ext = os.path.splitext(file_path)[1]
    if ext == ".toml":
        reason = _check_toml(file_path)
    elif ext == ".py":
        reason = _check_python(file_path)
    elif ext in _EXT_TO_LANG:
        reason = _check_regex_lang(file_path, _EXT_TO_LANG[ext])

    if reason:
        json.dump({"decision": "block", "reason": reason}, sys.stdout)


if __name__ == "__main__":
    main()

File: hooks-rs/src/hooks/block_scrape_interval.rs
use crate::input::HookInput;
use crate::output::block;
use crate::python_ast::{LineIndex, dotted_name, is_numeric_constant, parse_suite, walk_suite};
use regex::Regex;
use rustpython_parser::ast;
use std::fs;
use std::path::Path;
use toml::Value;

const MIN_INTERVAL: f64 = 1.0;

type SleepPattern = (&'static str, &'static str);

struct LangDef {
    http_re: &'static str,
    sleep_patterns: &'static [SleepPattern],
}

const RUST_SLEEP_PATTERNS: &[SleepPattern] = &[
    (r"Duration::from_secs\(\s*(\d+)", "s"),
    (r"Duration::from_secs_f(?:32|64)\(\s*([\d.]+)", "s"),
    (r"Duration::from_millis\(\s*(\d+)", "ms"),
];

const JS_TS_SLEEP_PATTERNS: &[SleepPattern] = &[
    (r"setTimeout\([^,]+,\s*(\d+)", "ms"),
    (r"(?:sleep|delay)\(\s*(\d+)", "ms"),
    (r"waitForTimeout\(\s*(\d+)", "ms"),
];

const DART_SLEEP_PATTERNS: &[SleepPattern] = &[
    (r"Duration\(\s*seconds:\s*(\d+)", "s"),
    (r"Duration\(\s*milliseconds:\s*(\d+)", "ms"),
];

const PYTHON_HTTP_RE: &str = r"\b(?:requests|httpx|session|client)\.(?:get|post|put|patch|delete|head|options|request|send)\b|\baiohttp\.ClientSession\b|\bpage\.goto\b|\burllib\.request\.urlopen\b";
const NOQA_RE: &str = r"(?:#|//)\s*noqa:\s*scrape-interval";

pub fn run(input: &HookInput) {
    let file_path = input.tool_input.file_path_resolved();
    if file_path.is_empty() || !is_scrape_file(file_path) || is_test_file(file_path) {
        return;
    }

    let path = Path::new(file_path);
    if !path.is_file() {
        return;
    }

    let reason = match path.extension().and_then(|s| s.to_str()).unwrap_or("") {
        "toml" => check_toml(path),
        "py" => check_python(path),
        "rs" => check_regex_lang(path, rust_lang()),
        "js" | "ts" => check_regex_lang(path, js_ts_lang()),
        "dart" => check_regex_lang(path, dart_lang()),
        _ => None,
    };

    if let Some(reason) = reason {
        block(&reason);
    }
}

fn rust_lang() -> LangDef {
    LangDef {
        http_re: r"\breqwest::(?:get|Client)\b|\bclient\.(?:get|post|put|patch|delete|head|request)\(|\.send\(\)\.await\b|\bhyper::Client\b|\bsurf::(?:get|post|put|delete)\b",
        sleep_patterns: RUST_SLEEP_PATTERNS,
    }
}

fn js_ts_lang() -> LangDef {
    LangDef {
        http_re: r"\bfetch\(|\baxios\.(?:get|post|put|patch|delete|head|request)\(|\bgot\(|\bpage\.goto\(|\bky\.(?:get|post|put|patch|delete|head)\(",
        sleep_patterns: JS_TS_SLEEP_PATTERNS,
    }
}

fn dart_lang() -> LangDef {
    LangDef {
        http_re: r"\bhttp\.(?:get|post|put|patch|delete|head|read)\(|\bclient\.(?:get|post|put|patch|delete|head|send)\(|\bDio\(\)",
        sleep_patterns: DART_SLEEP_PATTERNS,
    }
}

fn check_toml(path: &Path) -> Option<String> {
    let basename = basename(path);
    let source = fs::read_to_string(path).ok()?;
    let data: Value = match toml::from_str(&source) {
        Ok(value) => value,
        Err(err) => {
            return Some(format!("{basename}: TOML を解析できません: {err}"));
        }
    };

    let Some(interval_value) = find_interval(&data) else {
        return Some(format!(
            "{basename}: interval キーが定義されていません。\nscrape設定には interval >= {MIN_INTERVAL}s を含めてください。"
        ));
    };
    let Some(seconds) = resolve_interval(interval_value) else {
        return Some(format!(
            "{basename}: interval の値を解析できません: {}",
            display_toml_value(interval_value)
        ));
    };
    if seconds < MIN_INTERVAL {
        return Some(format!(
            "{basename}: interval = {} ({seconds}s) は最低 {MIN_INTERVAL}s 未満です。",
            display_toml_value(interval_value)
        ));
    }
    None
}

fn find_interval(value: &Value) -> Option<&Value> {
    let Value::Table(table) = value else {
        return None;
    };
    if let Some(interval) = table.get("interval") {
        return Some(interval);
    }
    table.values().find_map(find_interval)
}

fn resolve_interval(value: &Value) -> Option<f64> {
    match value {
        Value::Integer(value) => Some(*value as f64),
        Value::Float(value) => Some(*value),
        Value::String(value) => parse_interval_str(value),
        _ => None,
    }
}

fn parse_interval_str(value: &str) -> Option<f64> {
    let value = value.trim();
    if let Some(ms) = value.strip_suffix("ms") {
        return ms.parse::<f64>().ok().map(|v| v / 1000.0);
    }
    if let Some(sec) = value.strip_suffix('s') {
        return sec.parse::<f64>().ok();
    }
    value.parse::<f64>().ok()
}

fn display_toml_value(value: &Value) -> String {
    match value {
        Value::String(value) => format!("{value:?}"),
        _ => value.to_string(),
    }
}

fn check_python(path: &Path) -> Option<String> {
    let source = fs::read_to_string(path).ok()?;
    let suite = parse_suite(&source, &path.display().to_string())?;
    let index = LineIndex::new(&source);
    let noqa_re = Regex::new(NOQA_RE).ok()?;
    let mut low_sleeps = Vec::new();
    let mut has_adequate_sleep = false;

    walk_suite(
        &suite,
        &mut |_| {},
        &mut |expr| {
            let ast::Expr::Call(call) = expr else {
                return;
            };
            let Some(func_name) = sleep_func_name(&call.func) else {
                return;
            };
            let Some(arg) = call.args.first() else {
                return;
            };
            let Some((value, _)) = is_numeric_constant(arg) else {
                return;
            };
            if value >= MIN_INTERVAL {
                has_adequate_sleep = true;
                return;
            }
            let line = index.line_for(expr);
            if noqa_re.is_match(index.line_text(line)) {
                return;
            }
            low_sleeps.push((line, func_name.to_string(), value));
        },
        &mut |_| {},
    );

    let basename = basename(path);
    if !low_sleeps.is_empty() {
        let details = low_sleeps
            .iter()
            .map(|(line, call, value)| format!("  L{line}: {call}({value})"))
            .collect::<Vec<_>>()
            .join("\n");
        return Some(format!(
            "{basename}: sleep の値が {MIN_INTERVAL}s 未満です。\n{details}\nリクエスト間隔を {MIN_INTERVAL}s 以上にしてください (# noqa: scrape-interval で除外可)。"
        ));
    }

    if has_http_regex(&source, PYTHON_HTTP_RE, &noqa_re) && !has_adequate_sleep {
        return Some(format!(
            "{basename}: HTTP呼び出しがありますが sleep >= {MIN_INTERVAL}s が見つかりません。\nリクエスト間隔を確保する sleep を追加してください (# noqa: scrape-interval で除外可)。"
        ));
    }
    None
}

fn sleep_func_name(func: &ast::Expr) -> Option<&'static str> {
    match dotted_name(func).as_deref() {
        Some("time.sleep") => Some("time.sleep"),
        Some("asyncio.sleep") => Some("asyncio.sleep"),
        Some("sleep") => Some("sleep"),
        _ => None,
    }
}

fn check_regex_lang(path: &Path, lang: LangDef) -> Option<String> {
    let source = fs::read_to_string(path).ok()?;
    let lines: Vec<&str> = source.lines().collect();
    let noqa_re = Regex::new(NOQA_RE).ok()?;
    let sleep_patterns = compile_sleep_patterns(lang.sleep_patterns);
    let basename = basename(path);

    let low_sleeps = find_low_sleeps_regex(&lines, &sleep_patterns, &noqa_re);
    if !low_sleeps.is_empty() {
        let details = low_sleeps
            .iter()
            .map(|(line, text)| format!("  L{line}: {text}"))
            .collect::<Vec<_>>()
            .join("\n");
        return Some(format!(
            "{basename}: sleep/delay の値が {MIN_INTERVAL}s 未満です。\n{details}\nリクエスト間隔を {MIN_INTERVAL}s 以上にしてください (// noqa: scrape-interval で除外可)。"
        ));
    }

    if has_http_regex(&source, lang.http_re, &noqa_re)
        && !has_adequate_sleep_regex(&lines, &sleep_patterns)
    {
        return Some(format!(
            "{basename}: HTTP呼び出しがありますが sleep >= {MIN_INTERVAL}s が見つかりません。\nリクエスト間隔を確保する sleep/delay を追加してください (// noqa: scrape-interval で除外可)。"
        ));
    }
    None
}

fn compile_sleep_patterns(patterns: &[SleepPattern]) -> Vec<(Regex, &'static str)> {
    patterns
        .iter()
        .filter_map(|(pattern, unit)| Regex::new(pattern).ok().map(|re| (re, *unit)))
        .collect()
}

fn find_low_sleeps_regex(
    lines: &[&str],
    patterns: &[(Regex, &'static str)],
    noqa_re: &Regex,
) -> Vec<(usize, String)> {
    let mut violations = Vec::new();
    for (idx, line) in lines.iter().enumerate() {
        if noqa_re.is_match(line) {
            continue;
        }
        for (pattern, unit) in patterns {
            let Some(caps) = pattern.captures(line) else {
                continue;
            };
            let Some(value) = caps.get(1).and_then(|m| m.as_str().parse::<f64>().ok()) else {
                continue;
            };
            let seconds = if *unit == "ms" { value / 1000.0 } else { value };
            if seconds < MIN_INTERVAL {
                violations.push((idx + 1, line.trim().to_string()));
                break;
            }
        }
    }
    violations
}

fn has_http_regex(source: &str, pattern: &str, noqa_re: &Regex) -> bool {
    let Ok(http_re) = Regex::new(pattern) else {
        return false;
    };
    source
        .lines()
        .any(|line| !noqa_re.is_match(line) && http_re.is_match(line))
}

fn has_adequate_sleep_regex(lines: &[&str], patterns: &[(Regex, &'static str)]) -> bool {
    lines.iter().any(|line| {
        patterns.iter().any(|(pattern, unit)| {
            pattern
                .captures(line)
                .and_then(|caps| caps.get(1).and_then(|m| m.as_str().parse::<f64>().ok()))
                .is_some_and(|value| {
                    let seconds = if *unit == "ms" { value / 1000.0 } else { value };
                    seconds >= MIN_INTERVAL
                })
        })
    })
}

fn is_scrape_file(file_path: &str) -> bool {
    let normalized = file_path.replace('\\', "/");
    normalized.contains("/scrape/") || normalized.starts_with("scrape/")
}

fn is_test_file(file_path: &str) -> bool {
    let normalized = file_path.replace('\\', "/");
    normalized.contains("/tests/")
        || Path::new(file_path)
            .file_name()
            .and_then(|s| s.to_str())
            .is_some_and(|name| name.starts_with("test_"))
}

fn basename(path: &Path) -> String {
    path.file_name()
        .and_then(|s| s.to_str())
        .unwrap_or_default()
        .to_string()
}
